File: aoc24c/day19.c
// Advent of Code 2024
// Dr Lee A. Christie
//
// GitHub:   @leechristie
// Mastodon: @0x1ac@techhub.social
// Website:  leechristie.com

#include <ctype.h>
#include <stdio.h>

#include "strings.h"
#include "timing.h"

#define MAX_PATTERN_LENGTH (8)
#define MAX_PATTERNS (500)
#define MAX_TARGET_LENGTH (60)

static char patterns[MAX_PATTERNS + 1][MAX_PATTERN_LENGTH + 1];
static StringBuffer target;

static void read_patterns(FILE * const file) {
    int pattern_index = 0;
    int character_index = 0;
    int c = getc(file);
    while (true) {
        if (c == ',' || c == '\n') {
            patterns[pattern_index][character_index] = '\0';
            if (c == '\n') {
                patterns[pattern_index + 1][0] = '\0';
                break;
            }
            pattern_index++;
            character_index = 0;
        } else if (c == ' ') {
            // nothing
        } else {
            assert(pattern_index < MAX_PATTERNS);
            assert(character_index < MAX_PATTERN_LENGTH);
            assert(isalpha(c));
            patterns[pattern_index][character_index] = (char) c;
            character_index++;
        }
        c = getc(file);
    }
}

bool read_line(FILE * const file, StringBuffer * const buffer) {
    StringBuffer_Clear(buffer);
    while (true) {
        const int c = getc(file);
        if (c == '\n' || c == EOF) {
            return buffer->length > 0;
        }
        StringBuffer_Append(buffer, (char) c);
    }
}

static long long memoized_result[MAX_TARGET_LENGTH] = {-1};

static size_t fits_here(const size_t next_match_index, const int lower) {
    size_t index = 0;
    char symbol = patterns[next_match_index][index];
    while (symbol != '\0') {
        const char match_symbol = target.data[index + lower];
        if (match_symbol == '\0') {
            return 0;
        }
        if (match_symbol != symbol) {
            return 0;
        }
        index++;
        symbol = patterns[next_match_index][index];
    }
    return index;
}

static long long ways_of_tail_matching(const size_t lower) {
    assert(lower < MAX_TARGET_LENGTH);
    if (memoized_result[lower] != -1) {
        return memoized_result[lower];
    }
    if (target.data[lower] == '\0') {
        return 1;
    }
    size_t found_match_next_tails[MAX_PATTERNS + 1] = {0};
    size_t num_matches = 0;
    size_t next_match_index = 0;
    while (patterns[next_match_index][0] != '\0') {
        const size_t offset = fits_here(next_match_index, lower);
        if (offset) {
            found_match_next_tails[num_matches++] = lower + offset;
        }
        next_match_index++;
    }
    if (num_matches == 0) {
        return 0;
    }
    long long rv = 0;
    for (size_t i = 0; i < num_matches; i++) {
        assert(found_match_next_tails[i] > 0);
        rv += ways_of_tail_matching(found_match_next_tails[i]);
    }
    memoized_result[lower] = rv;
    return rv;
}

int day19() {

    start_timer();

    int part1 = 0;
    long long part2 = 0;

    FILE * file = fopen("input19.txt", "r");
    if (file == NULL) {
        printf("unable to open input file");
        return 1;
    }

    // read the patterns
    read_patterns(file);

    // read blank line
    getc(file);

    // for each target
    StringBuffer_Init(&target);
    while (read_line(file, &target)) {
        assert(target.length <= MAX_TARGET_LENGTH);
        for (size_t i = 0; i < MAX_TARGET_LENGTH; i++) {
            memoized_result[i] = -1;
        }
        long long ways = ways_of_tail_matching(0);
        if (ways > 0) {
            part1++;
        }
        part2 += ways;
    }
    StringBuffer_DeInit(&target);

    fclose(file);

    const double time = stop_timer();
    printf("Advent of Code 2024\n");
    printf("Day 19 - Linen Layout\n");
    printf("Part 1: %d\n", part1);
    printf("Part 2: %lld\n", part2);
    printf("Time Taken: %.6f s\n", time);
    return 0;

}

File: aoc24py/day19.py
# Advent of Code 2024
# Dr Lee A. Christie
#
# GitHub:   @leechristie
# Mastodon: @0x1ac@techhub.social
# Website:  leechristie.com
import re
import time
from functools import cache
from itertools import count


def read_input(filename: str) -> tuple[list[str], list[str]]:
    targets: list[str] = []
    with open(filename) as file:
        materials = next(file).strip().split(', ')
        assert next(file).strip() == ''
        for target in file:
            target = target.strip()
            targets.append(target)
    return materials, targets


def symbol_set(materials: list[str]) -> set[str]:
    rv: set[str] = set()
    for material in materials:
        for symbol in material:
            rv.add(symbol)
    return rv


def encode(target: str, symbols: dict[str, int]) -> list[int]:
    rv: list[int] = []
    for character in target:
        rv.append(symbols[character])
    return rv


def fits_here(next_match: list[int], encoded_target: list[int], lower: int) -> bool:
    for index, symbol in enumerate(next_match):
        offset_index = index + lower
        if offset_index >= len(encoded_target):
            return False
        if symbol != encoded_target[offset_index]:
            return False
    return True


def ways_of_matching(encoded_target: list[int], encoded_materials: list[list[int]]) -> int:

    @cache
    def ways_of_tail_matching(lower: int) -> int:

        if lower == len(encoded_target):
            return 1

        found_match_next_tails: list[int] = []
        for next_match in encoded_materials:
            if fits_here(next_match, encoded_target, lower):
                found_match_next_tails.append(lower + len(next_match))

        if not found_match_next_tails:
            return 0

        rv = 0
        for next_lower in found_match_next_tails:
            rv += ways_of_tail_matching(next_lower)
        return rv

    return ways_of_tail_matching(0)


def day19() -> None:

    start: float = time.perf_counter()

    materials, targets = read_input('input19.txt')

    symbols = {s: i for s, i in zip(sorted(symbol_set(materials)), count())}

    encoded_materials: list[list[int]] = [encode(material, symbols) for material in materials]

    part1: int = 0
    part2: int = 0
    for target in targets:
        encoded_target: list[int] = encode(target, symbols)
        number_of_ways = ways_of_matching(encoded_target, encoded_materials)
        if number_of_ways:
            part1 += 1
        part2 += number_of_ways

    stop: float = time.perf_counter()

    print("Advent of Code 2024")
    print("Day 19 - Linen Layout")
    print(f"Part 1: {part1}")
    print(f"Part 2: {part2}")
    print(f"Time Taken: {stop-start:.6f} s")
